crates/ruma-html/src/html/matrix.rs

```diff
@@ -0,0 +1,721 @@
+/// and do not use the [`MatrixElement::H`] variant.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct HeadingLevel(u8);
+
+impl HeadingLevel {
+    /// The value of the level.
+    ///
+    /// Can only be a value between 1 and 6 included.
+    pub fn value(&self) -> u8 {
+        self.0
+    }
+}
+
+impl PartialEq<u8> for HeadingLevel {
+    fn eq(&self, other: &u8) -> bool {
+        self.0.eq(other)
+    }
+}
+
+/// The supported data of a `<a>` HTML element.
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct AnchorData {
+    /// The name of the anchor.
+    pub name: Option<StrTendril>,
+
+    /// Where to display the linked URL.
+    pub target: Option<StrTendril>,
+
+    /// The URL that the hyperlink points to.
+    pub href: Option<AnchorUri>,
+}
+
+impl AnchorData {
+    /// Construct an empty `AnchorData`.
+    fn new() -> Self {
+        Self { name: None, target: None, href: None }
+    }
+
+    /// Parse the given attributes to construct a new `AnchorData`.
+    ///
+    /// Returns a tuple containing the constructed data and the remaining unsupported attributes.
+    #[allow(clippy::mutable_key_type)]
+    fn parse(attrs: &BTreeSet<Attribute>) -> (Self, BTreeSet<Attribute>) {
+        let mut data = Self::new();
+        let mut remaining_attrs = BTreeSet::new();
+
+        for attr in attrs {
+            if attr.name.ns != ns!() {
+                remaining_attrs.insert(attr.clone());
+                continue;
+            }
+
+            match attr.name.local.as_bytes() {
+                b"name" => {
+                    data.name = Some(attr.value.clone());
+                }
+                b"target" => {
+                    data.target = Some(attr.value.clone());
+                }
+                b"href" => {
+                    if let Some(uri) = AnchorUri::parse(&attr.value) {
+                        data.href = Some(uri);
+                    } else {
+                        remaining_attrs.insert(attr.clone());
+                    }
+                }
+                _ => {
+                    remaining_attrs.insert(attr.clone());
+                }
+            }
+        }
+
+        (data, remaining_attrs)
+    }
+}
+
+/// A URI as a value for the `href` attribute of a `<a>` HTML element.
+///
+/// This is a helper type that recognizes `matrix:` and `https://matrix.to` URIs to detect mentions.
+///
+/// If the URI is an invalid Matrix URI or does not use one of the suggested schemes, the `href`
+/// attribute will be in the `attrs` list of [`MatrixElementData`].
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum AnchorUri {
+    /// A `matrix:` URI.
+    Matrix(MatrixUri),
+    /// A `https://matrix.to` URI.
+    MatrixTo(MatrixToUri),
+    /// An other URL using one of the suggested schemes.
+    ///
+    /// Those schemes are:
+    ///
+    /// * `https`
+    /// * `http`
+    /// * `ftp`
+    /// * `mailto`
+    /// * `magnet`
+    Other(StrTendril),
```

Comment:
You may want to use `bstr` (https://docs.rs/bstr) instead of tendril. I believe it's a bit lighter and may fulfill your needs.

Reply:
Oh, I notice that html5ever is using tendril, so forget about my comment.